design/keycloak.x/configuration.md

```diff
@@ -0,0 +1,159 @@
+
+* High-level property - for example `db.vendor=oracle12` will behind the scenes set `quarkus.datasource.jdbc.driver` and
+  `quarkus.hibernate-orm.dialect`
+* Mapping/alias - for example `http.ssl.certificate.file` will map to `quarkus.http.ssl.certificate.file`
+
+We will also allow setting Quarkus properties directly. This will not be recommended approach, but available as an option
+to power-users.
+
+## Documenting Properties
+
+We want to document all all properties that a user can set (this does not include the low-level Quarkus properties). We
+would like to achieve something similar to https://quarkus.io/guides/all-config.
+
+In the documentation the property key used in the properties file should be used. It should include a description, the 
+type and the default value.
+
+We would also like the ability to list properties on the command-line by running `bin/kc.sh --list-config-options`.
+
+For SPIs and Providers we should extend the ProviderFactory to add an option method to return information on what
+properties it supports.
+
+## Config Options
+
+The aim is to start small with what config options we provide, and extend when there is a need. Great care should be 
+taken in reducing the amount of configuration needed, where opinionated/best-practice behaviour should be favoured, with
+high-level and simple configuration options introduced only when needed.
+
+Below is a list of config by category we should introduce initially:
+
+### HTTP
+
+Keycloak.X will not come with the capability to generate a self-signed certificate. Instead, Keycloak by default will
+complain if a certificate has not been configured, but will allow explicitly disabling SSL for development purposes.
+
+* http.mode - Options `ssl` (default), `unsecure`  
+* http.port
+* http.ssl-port
+* http.ssl.certificate.file
+* http.ssl.certificate.key-file
+* http.ssl.certificate.key-store-file
+* http.ssl.certificate.key-store-file-type
+* http.ssl.certificate.key-store-password
+* http.ssl.certificate.trust-store-file
+* http.ssl.certificate.trust-store-file-type
+* http.ssl.certificate.trust-store-password
+* http.ssl.cipher-suites
+* http.ssl.protocols
+* http.ssl.client-auth
+
+### Proxy Mode
+
+The proxy mode will allow an easy way to configure if Keycloak is hosted behind a reverse-proxy or not, including how
+SSL is terminated.
+
+* proxy.mode - Options are `none` (default), `reencrypt`, `edge` and `passthrough`.
+
+Setting `proxy.mode` will under the covers configure `http.mode` as well as `quarkus.http.proxy-address-forwarding`.
+
+## Database
+
+The aim is that a user should not need to understand Hibernate, JDBC drivers, etc. to be able to configure the database.
+As such we should introduce a high-level option `db.vendor` instead of requiring users to set JDBC driver and Hibernate
+dialects.
+
+We will support either `db.url` which will support a JDBC URL, or using separate properties to build-up the URL behind the
```

Comment:
Does "either" mean it is not decided yet or "both"? If only construction is possible, it should also be supported to add custom parameters to the URL since there are so many different parameters for the different databases.

Reply:
Reading this a second time, it probably means both. Then I would really like to have a db.custom_params configuration property.

Reply:
The thinking was to support both. Big question really is what is most convenient, also it is slightly tricky to let you "build" the JDBC URL in an intuitive way since there isn't much of a standard there AFAIK, where all vendors have different syntax. So perhaps the "builder" approach is for novice users, while the direct URL approach is for power users.

Reply:
Thinking more about it, I think we should not use abbreviations but `database`. It looks weird to me and not so friendly if we just use `db`.

It is also aligned with other examples where the correct name (e.g.: certificate and not cert|crt) is being used.

There are also some other properties that we may want to expose here but we can always document them later, like enable metrics.